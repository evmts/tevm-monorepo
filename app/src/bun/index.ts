import { BrowserWindow } from "electrobun/bun";

const win = new BrowserWindow({
  title: "Hello Electrobun",
  url: "views://ui/index.html",
});
