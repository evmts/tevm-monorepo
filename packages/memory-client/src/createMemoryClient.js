import { createBaseClient } from '@tevm/base-client'
import {
	eip1993EventEmitter,
	ethActions,
	requestEip1193,
	tevmActions,
	tevmSend,
} from '@tevm/decorators'

/**
 * A local EVM instance running in JavaScript. Similar to Anvil in your browser
 * @param {import('@tevm/base-client').BaseClientOptions} [options]
 * @returns {import('./MemoryClient.js').MemoryClient}
 * @example
 * ```ts
 * import { createMemoryClient } from "tevm"
 * import { MyERC721 } from './MyERC721.sol'
 *
 * const tevm = createMemoryClient({
 * 	fork: {
 * 	  url: "https://mainnet.optimism.io",
 * 	},
 * })
 *
 * const address = '0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045',

 * await tevm.runContractCall(
 *   MyERC721.write.mint({
 *     caller: address,
 *   }),
 * )
 *
 * const balance = await tevm.runContractCall(
 *  MyERC721.read.balanceOf({
 *  caller: address,
 *  }),
 *  )
 *  console.log(balance) // 1n
 *  ```
 */
export const createMemoryClient = (options) => {
	return createBaseClient(options)
		.extend(tevmSend())
		.extend(eip1993EventEmitter())
		.extend(requestEip1193())
		.extend(ethActions())
		.extend(tevmActions())
}
