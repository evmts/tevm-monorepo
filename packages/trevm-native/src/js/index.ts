import { createLogger } from '@tevm/logger';
import { InternalError } from '@tevm/errors';
import type { StateManager } from '@tevm/state';
import type { Common } from '@tevm/common';
import type { Chain } from '@tevm/blockchain';

// Import native module
import { init_trevm, TrevmWrapper } from '../../dist/index.node';

export interface TrevmEvmOptions {
  common: Common;
  stateManager: StateManager;
  blockchain: Chain;
  customPrecompiles?: any[];
  profiler?: boolean;
  allowUnlimitedContractSize?: boolean;
  loggingLevel?: string;
}

export interface TrevmRunCallOptions {
  caller?: string;
  to?: string;
  data?: string;
  value?: bigint;
  gasLimit?: bigint;
  skipBalance?: boolean;
}

export interface TrevmRunCallResult {
  returnValue: string;
  executionGasUsed: string;
  error?: string;
}

export interface TrevmEvm {
  runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult>;
}

/**
 * Creates a trevm-based EVM instance
 * @param options EVM options
 * @returns A trevm EVM instance
 */
export const createTrevm = async ({
  common,
  stateManager,
  blockchain,
  customPrecompiles,
  profiler,
  allowUnlimitedContractSize,
  loggingLevel,
}: TrevmEvmOptions): Promise<TrevmEvm> => {
  const logger = createLogger({
    name: '@tevm/trevm',
    level: loggingLevel ?? 'warn',
  });
  
  logger.debug({
    allowUnlimitedContractSize,
    profiler,
    customPrecompiles: customPrecompiles?.map(c => c.address.toString()),
  });

  try {
    // Initialize the native trevm module with callbacks to the StateManager
    const nativeTrevm = await init_trevm({
      chainId: common.chainId(),
      allowUnlimitedContractSize: allowUnlimitedContractSize ?? false,
      stateCallbacks: {
        getAccount: async (address: string) => {
          try {
            const account = await stateManager.getAccount(address);
            if (!account) return null;
            
            return {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? `0x${Buffer.from(account.codeHash).toString('hex')}` : undefined,
              code: account.code ? `0x${Buffer.from(account.code).toString('hex')}` : undefined,
            };
          } catch (error) {
            logger.error(`Error in getAccount for ${address}:`, error);
            return null;
          }
        },
        putAccount: async (address: string, account: any) => {
          try {
            await stateManager.putAccount(address, {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? Buffer.from(account.codeHash.slice(2), 'hex') : undefined,
              code: account.code ? Buffer.from(account.code.slice(2), 'hex') : undefined,
            });
            return true;
          } catch (error) {
            logger.error(`Error in putAccount for ${address}:`, error);
            return false;
          }
        },
        getStorage: async (address: string, key: string) => {
          try {
            const value = await stateManager.getContractStorage(address, key);
            return value ? `0x${Buffer.from(value).toString('hex')}` : '0x0';
          } catch (error) {
            logger.error(`Error in getStorage for ${address} key ${key}:`, error);
            return '0x0';
          }
        },
        putStorage: async (address: string, key: string, value: string) => {
          try {
            await stateManager.putContractStorage(
              address,
              key,
              Buffer.from(value.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putStorage for ${address} key ${key}:`, error);
            return false;
          }
        },
        getCode: async (address: string) => {
          try {
            const code = await stateManager.getContractCode(address);
            return code ? `0x${Buffer.from(code).toString('hex')}` : '0x';
          } catch (error) {
            logger.error(`Error in getCode for ${address}:`, error);
            return '0x';
          }
        },
        putCode: async (address: string, code: string) => {
          try {
            await stateManager.putContractCode(
              address,
              Buffer.from(code.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putCode for ${address}:`, error);
            return false;
          }
        },
        getBlockHash: async (blockNumber: number) => {
          try {
            const block = await blockchain.getBlock(BigInt(blockNumber));
            return block ? `0x${Buffer.from(block.header.hash()).toString('hex')}` : null;
          } catch (error) {
            logger.error(`Error in getBlockHash for block ${blockNumber}:`, error);
            return null;
          }
        }
      }
    });

    // Return a JavaScript wrapper around the native trevm that matches the Evm interface
    return {
      async runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult> {
        try {
          return await nativeTrevm.run_call({
            caller: opts.caller,
            to: opts.to,
            data: opts.data,
            value: opts.value,
            gas_limit: opts.gasLimit,
            skip_balance: opts.skipBalance,
          });
        } catch (error) {
          logger.error('Error in runCall:', error);
          throw new InternalError(`Trevm execution error: ${error}`);
        }
      }
    };
  } catch (error) {
    logger.error('Error initializing trevm:', error);
    throw new InternalError(`Failed to initialize trevm: ${error}`);
  }
};