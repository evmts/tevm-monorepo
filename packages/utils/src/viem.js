// All these are needed to use tevm and there is no reason to reinvent the wheel on these viem utils
export { formatAbi, parseAbi } from 'abitype'
export { mnemonicToAccount } from 'viem/accounts'
export {
	toRlp,
	parseGwei,
	parseEther,
	bytesToHex,
	hexToBool,
	hexToBytes,
	hexToBigInt,
	hexToNumber,
	numberToHex,
	boolToHex,
	stringToHex,
	fromHex,
	fromBytes,
	toBytes,
	toHex,
	encodePacked,
	encodeDeployData,
	encodeErrorResult,
	encodeEventTopics,
	encodeAbiParameters,
	encodeFunctionData,
	encodeFunctionResult,
	decodeFunctionData,
	decodeFunctionResult,
	decodeEventLog,
	decodeErrorResult,
	decodeAbiParameters,
	formatGwei,
	formatLog,
	formatEther,
	fromRlp,
	getAddress,
	isAddress,
	isBytes,
	isHex,
	keccak256,
	boolToBytes,
	bytesToBool,
	hexToString,
	bytesToBigint,
	bytesToBigInt,
	bytesToNumber,
} from 'viem/utils'
