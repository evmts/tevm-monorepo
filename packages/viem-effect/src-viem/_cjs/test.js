'use strict'
Object.defineProperty(exports, '__esModule', { value: true })
exports.stopImpersonatingAccount =
	exports.setRpcUrl =
	exports.snapshot =
	exports.setStorageAt =
	exports.setNonce =
	exports.setNextBlockTimestamp =
	exports.setNextBlockBaseFeePerGas =
	exports.setMinGasPrice =
	exports.setLoggingEnabled =
	exports.setIntervalMining =
	exports.setCoinbase =
	exports.setCode =
	exports.setBlockTimestampInterval =
	exports.setBlockGasLimit =
	exports.setBalance =
	exports.setAutomine =
	exports.sendUnsignedTransaction =
	exports.revert =
	exports.reset =
	exports.removeBlockTimestampInterval =
	exports.mine =
	exports.inspectTxpool =
	exports.increaseTime =
	exports.impersonateAccount =
	exports.getTxpoolStatus =
	exports.getTxpoolContent =
	exports.getAutomine =
	exports.dropTransaction =
		void 0
const dropTransaction_js_1 = require('./actions/test/dropTransaction.js')
Object.defineProperty(exports, 'dropTransaction', {
	enumerable: true,
	get: function () {
		return dropTransaction_js_1.dropTransaction
	},
})
const getAutomine_js_1 = require('./actions/test/getAutomine.js')
Object.defineProperty(exports, 'getAutomine', {
	enumerable: true,
	get: function () {
		return getAutomine_js_1.getAutomine
	},
})
const getTxpoolContent_js_1 = require('./actions/test/getTxpoolContent.js')
Object.defineProperty(exports, 'getTxpoolContent', {
	enumerable: true,
	get: function () {
		return getTxpoolContent_js_1.getTxpoolContent
	},
})
const getTxpoolStatus_js_1 = require('./actions/test/getTxpoolStatus.js')
Object.defineProperty(exports, 'getTxpoolStatus', {
	enumerable: true,
	get: function () {
		return getTxpoolStatus_js_1.getTxpoolStatus
	},
})
const impersonateAccount_js_1 = require('./actions/test/impersonateAccount.js')
Object.defineProperty(exports, 'impersonateAccount', {
	enumerable: true,
	get: function () {
		return impersonateAccount_js_1.impersonateAccount
	},
})
const increaseTime_js_1 = require('./actions/test/increaseTime.js')
Object.defineProperty(exports, 'increaseTime', {
	enumerable: true,
	get: function () {
		return increaseTime_js_1.increaseTime
	},
})
const inspectTxpool_js_1 = require('./actions/test/inspectTxpool.js')
Object.defineProperty(exports, 'inspectTxpool', {
	enumerable: true,
	get: function () {
		return inspectTxpool_js_1.inspectTxpool
	},
})
const mine_js_1 = require('./actions/test/mine.js')
Object.defineProperty(exports, 'mine', {
	enumerable: true,
	get: function () {
		return mine_js_1.mine
	},
})
const removeBlockTimestampInterval_js_1 = require('./actions/test/removeBlockTimestampInterval.js')
Object.defineProperty(exports, 'removeBlockTimestampInterval', {
	enumerable: true,
	get: function () {
		return removeBlockTimestampInterval_js_1.removeBlockTimestampInterval
	},
})
const reset_js_1 = require('./actions/test/reset.js')
Object.defineProperty(exports, 'reset', {
	enumerable: true,
	get: function () {
		return reset_js_1.reset
	},
})
const revert_js_1 = require('./actions/test/revert.js')
Object.defineProperty(exports, 'revert', {
	enumerable: true,
	get: function () {
		return revert_js_1.revert
	},
})
const sendUnsignedTransaction_js_1 = require('./actions/test/sendUnsignedTransaction.js')
Object.defineProperty(exports, 'sendUnsignedTransaction', {
	enumerable: true,
	get: function () {
		return sendUnsignedTransaction_js_1.sendUnsignedTransaction
	},
})
const setAutomine_js_1 = require('./actions/test/setAutomine.js')
Object.defineProperty(exports, 'setAutomine', {
	enumerable: true,
	get: function () {
		return setAutomine_js_1.setAutomine
	},
})
const setBalance_js_1 = require('./actions/test/setBalance.js')
Object.defineProperty(exports, 'setBalance', {
	enumerable: true,
	get: function () {
		return setBalance_js_1.setBalance
	},
})
const setBlockGasLimit_js_1 = require('./actions/test/setBlockGasLimit.js')
Object.defineProperty(exports, 'setBlockGasLimit', {
	enumerable: true,
	get: function () {
		return setBlockGasLimit_js_1.setBlockGasLimit
	},
})
const setBlockTimestampInterval_js_1 = require('./actions/test/setBlockTimestampInterval.js')
Object.defineProperty(exports, 'setBlockTimestampInterval', {
	enumerable: true,
	get: function () {
		return setBlockTimestampInterval_js_1.setBlockTimestampInterval
	},
})
const setCode_js_1 = require('./actions/test/setCode.js')
Object.defineProperty(exports, 'setCode', {
	enumerable: true,
	get: function () {
		return setCode_js_1.setCode
	},
})
const setCoinbase_js_1 = require('./actions/test/setCoinbase.js')
Object.defineProperty(exports, 'setCoinbase', {
	enumerable: true,
	get: function () {
		return setCoinbase_js_1.setCoinbase
	},
})
const setIntervalMining_js_1 = require('./actions/test/setIntervalMining.js')
Object.defineProperty(exports, 'setIntervalMining', {
	enumerable: true,
	get: function () {
		return setIntervalMining_js_1.setIntervalMining
	},
})
const setLoggingEnabled_js_1 = require('./actions/test/setLoggingEnabled.js')
Object.defineProperty(exports, 'setLoggingEnabled', {
	enumerable: true,
	get: function () {
		return setLoggingEnabled_js_1.setLoggingEnabled
	},
})
const setMinGasPrice_js_1 = require('./actions/test/setMinGasPrice.js')
Object.defineProperty(exports, 'setMinGasPrice', {
	enumerable: true,
	get: function () {
		return setMinGasPrice_js_1.setMinGasPrice
	},
})
const setNextBlockBaseFeePerGas_js_1 = require('./actions/test/setNextBlockBaseFeePerGas.js')
Object.defineProperty(exports, 'setNextBlockBaseFeePerGas', {
	enumerable: true,
	get: function () {
		return setNextBlockBaseFeePerGas_js_1.setNextBlockBaseFeePerGas
	},
})
const setNextBlockTimestamp_js_1 = require('./actions/test/setNextBlockTimestamp.js')
Object.defineProperty(exports, 'setNextBlockTimestamp', {
	enumerable: true,
	get: function () {
		return setNextBlockTimestamp_js_1.setNextBlockTimestamp
	},
})
const setNonce_js_1 = require('./actions/test/setNonce.js')
Object.defineProperty(exports, 'setNonce', {
	enumerable: true,
	get: function () {
		return setNonce_js_1.setNonce
	},
})
const setStorageAt_js_1 = require('./actions/test/setStorageAt.js')
Object.defineProperty(exports, 'setStorageAt', {
	enumerable: true,
	get: function () {
		return setStorageAt_js_1.setStorageAt
	},
})
const snapshot_js_1 = require('./actions/test/snapshot.js')
Object.defineProperty(exports, 'snapshot', {
	enumerable: true,
	get: function () {
		return snapshot_js_1.snapshot
	},
})
const setRpcUrl_js_1 = require('./actions/test/setRpcUrl.js')
Object.defineProperty(exports, 'setRpcUrl', {
	enumerable: true,
	get: function () {
		return setRpcUrl_js_1.setRpcUrl
	},
})
const stopImpersonatingAccount_js_1 = require('./actions/test/stopImpersonatingAccount.js')
Object.defineProperty(exports, 'stopImpersonatingAccount', {
	enumerable: true,
	get: function () {
		return stopImpersonatingAccount_js_1.stopImpersonatingAccount
	},
})
//# sourceMappingURL=test.js.map
