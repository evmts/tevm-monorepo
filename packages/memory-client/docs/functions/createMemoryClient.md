**@tevm/memory-client** ∙ [README](../README.md) ∙ [API](../API.md)

***

[API](../API.md) > createMemoryClient

# Function: createMemoryClient()

> **createMemoryClient**(`options`?): `Promise`\<[`MemoryClient`](../type-aliases/MemoryClient.md)\>

A local EVM instance running in JavaScript. Similar to Anvil in your browser

## Parameters

▪ **options?**: [`CreateEVMOptions`](../type-aliases/CreateEVMOptions.md)= `{}`

## Returns

## Example

```ts
import { createMemoryClient } from "tevm"
import { createPublicClient, http } from "viem"
import { MyERC721 } from './MyERC721.sol'

const tevm = createMemoryClient({
	fork: {
	  url: "https://mainnet.optimism.io",
	},
})

const address = '0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045',

await tevm.runContractCall(
  MyERC721.write.mint({
    caller: address,
  }),
)

const balance = await tevm.runContractCall(
 MyERC721.read.balanceOf({
 caller: address,
 }),
 )
 console.log(balance) // 1n
 ```

## Source

[packages/memory-client/src/createMemoryClient.js:59](https://github.com/evmts/tevm-monorepo/blob/main/packages/memory-client/src/createMemoryClient.js#L59)

***
Generated using [typedoc-plugin-markdown](https://www.npmjs.com/package/typedoc-plugin-markdown) and [TypeDoc](https://typedoc.org/)
