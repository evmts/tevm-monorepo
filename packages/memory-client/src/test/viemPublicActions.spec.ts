import { beforeEach, describe, expect, it } from 'bun:test'
import { simpleContract } from '@tevm/test-utils'
import { type PublicActions, encodeFunctionData } from 'viem'
import type { MemoryClient } from '../MemoryClient.js'
import { createMemoryClient } from '../createMemoryClient.js'

describe('viemPublicActions', () => {
	let mc: MemoryClient
	let c = {
		simpleContract: simpleContract.withAddress(`0x${'00'.repeat(20)}`),
	}

	beforeEach(async () => {
		mc = createMemoryClient()
		const deployResult = await mc.tevmDeploy({
			bytecode: simpleContract.bytecode,
			abi: simpleContract.abi,
			args: [420n],
		})
		if (!deployResult.createdAddress) {
			throw new Error('contract never deployed')
		}
		c = {
			simpleContract: simpleContract.withAddress(deployResult.createdAddress),
		}
		await mc.tevmMine()
	})

	const tests: Record<keyof PublicActions, () => void> = {
		call: () => {
			it('should work', async () => {
				expect(
					await mc.call({
						to: c.simpleContract.address,
						data: encodeFunctionData(simpleContract.read.get()),
					}),
				).toEqual({
					data: '0x00000000000000000000000000000000000000000000000000000000000001a4',
				})
			})
		},
		createBlockFilter: () => {
			it.todo('not supported')
		},
		createContractEventFilter: () => {
			it.todo('not supported')
		},
		createEventFilter: () => {
			it.todo('not supported')
		},
		createPendingTransactionFilter: () => {
			it.todo('not supported')
		},
		estimateContractGas: () => {
			it('should work', async () => {
				expect(await mc.estimateContractGas(c.simpleContract.write.set(69n))).toBe(16771823n)
			})
		},
		estimateFeesPerGas: () => {},
		estimateGas: () => {},
		estimateMaxPriorityFeePerGas: () => {},
		getBalance: () => {},
		getBlobBaseFee: () => {},
		getBlock: () => {},
		getBlockNumber: () => {},
		getBlockTransactionCount: () => {},
		getBytecode: () => {},
		getChainId: () => {},
		getContractEvents: () => {},
		getEnsAddress: () => {},
		getEnsAvatar: () => {},
		getEnsName: () => {},
		getEnsResolver: () => {},
		getEnsText: () => {},
		getFeeHistory: () => {},
		getFilterChanges: () => {},
		getFilterLogs: () => {},
		getGasPrice: () => {},
		getLogs: () => {},
		getProof: () => {},
		getStorageAt: () => {},
		getTransaction: () => {},
		getTransactionConfirmations: () => {},
		getTransactionCount: () => {},
		getTransactionReceipt: () => {},
		multicall: () => {},
		prepareTransactionRequest: () => {},
		readContract: () => {},
		sendRawTransaction: () => {},
		simulateContract: () => {},
		uninstallFilter: () => {},
		verifyMessage: () => {},
		verifyTypedData: () => {},
		watchContractEvent: () => {},
		waitForTransactionReceipt: () => {},
		watchBlockNumber: () => {},
		watchBlocks: () => {},
		watchEvent: () => {},
		watchPendingTransactions: () => {},
	}

	Object.entries(tests).forEach(([actionName, actionTests]) => {
		describe(actionName, actionTests)
	})
})
