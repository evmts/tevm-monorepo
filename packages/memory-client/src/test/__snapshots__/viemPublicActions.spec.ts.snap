// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`viemPublicActions getBlock should work with blockHash 1`] = `
[
  {
    "accessList": [],
    "blobVersionedHashes": undefined,
    "blockHash": "redacted",
    "blockNumber": 1n,
    "chainId": undefined,
    "data": "0x608060405234801561000f575f80fd5b506040516101ef3803806101ef83398181016040528101906100319190610074565b805f819055505061009f565b5f80fd5b5f819050919050565b61005381610041565b811461005d575f80fd5b50565b5f8151905061006e8161004a565b92915050565b5f602082840312156100895761008861003d565b5b5f61009684828501610060565b91505092915050565b610143806100ac5f395ff3fe608060405234801561000f575f80fd5b5060043610610034575f3560e01c806360fe47b1146100385780636d4ce63c14610054575b5f80fd5b610052600480360381019061004d91906100ba565b610072565b005b61005c61007b565b60405161006991906100f4565b60405180910390f35b805f8190555050565b5f8054905090565b5f80fd5b5f819050919050565b61009981610087565b81146100a3575f80fd5b50565b5f813590506100b481610090565b92915050565b5f602082840312156100cf576100ce610083565b5b5f6100dc848285016100a6565b91505092915050565b6100ee81610087565b82525050565b5f6020820190506101075f8301846100e5565b9291505056fea264697066735822122019e943356c89506511b952171a3b4724d3152e5f4029bbb0ecc836d1365fcce464736f6c6343000816003300000000000000000000000000000000000000000000000000000000000001a4",
    "from": "0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266",
    "gas": 162455n,
    "gasPrice": 7n,
    "hash": "0x2a872fc2c05d90cbbdfbed7a5c831533dc1d02c1be4ab374b7d9c66e9ccec0e8",
    "maxFeePerGas": 7n,
    "maxPriorityFeePerGas": 0n,
    "nonce": 0,
    "to": null,
    "transactionIndex": 0,
    "type": "eip1559",
    "typeHex": "0x2",
    "v": undefined,
    "value": 0n,
    "yParity": undefined,
  },
]
`;

exports[`viemPublicActions getBlock should work with blockHash 2`] = `
{
  "baseFeePerGas": 7n,
  "blobGasUsed": 0n,
  "difficulty": 0n,
  "excessBlobGas": 0n,
  "extraData": "0x",
  "gasLimit": 30000000n,
  "gasUsed": 147721n,
  "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  "miner": "0x0000000000000000000000000000000000000000",
  "mixHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "nonce": "0x0000000000000000",
  "number": 1n,
  "parentBeaconBlockRoot": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "parentHash": "0x63b5bacdb65c8c9d10e77d77d40bda5b4dc010a4f7e02a10bdea21a2b1b0a454",
  "receiptsRoot": "0x73e9175f192532c39999c5b0e2a76d74343438183a65eee75c8b8913a92914a6",
  "requests": undefined,
  "requestsRoot": undefined,
  "sha3Uncles": "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
  "size": 2749n,
  "stateRoot": "0xfd597716848d3bd8c7525509bc25b611fde85624b582e1e7aa0dd5ce5d65fd48",
  "totalDifficulty": 0n,
  "transactionsRoot": "0xbf171948bc6a2ef34f1a272f9538e175d5b7a6f2b86b76b94cff5c2900ba8fc5",
  "uncles": [],
  "withdrawals": [],
  "withdrawalsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
}
`;

exports[`viemPublicActions getBlock should work with blocknumber 1`] = `[]`;

exports[`viemPublicActions getBlock should work with blocknumber 2`] = `
{
  "baseFeePerGas": 7n,
  "blobGasUsed": 0n,
  "difficulty": 17179869184n,
  "excessBlobGas": 0n,
  "extraData": "0x11bbe8db4e347b4e8c937c1c8370e4b5ed33adb3db69cbdb7a38e1e50b1b82fa",
  "gasLimit": 30000000n,
  "gasUsed": 0n,
  "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  "miner": "0x0000000000000000000000000000000000000000",
  "mixHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "nonce": "0x0000000000000042",
  "number": 0n,
  "parentBeaconBlockRoot": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "parentHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "receiptsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  "requests": undefined,
  "requestsRoot": undefined,
  "sha3Uncles": "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
  "size": 1597n,
  "stateRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  "totalDifficulty": 0n,
  "transactionsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
  "uncles": [],
  "withdrawals": [],
  "withdrawalsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
}
`;

exports[`viemPublicActions getTransaction should work 1`] = `
{
  "accessList": [],
  "blobVersionedHashes": undefined,
  "blockNumber": 1n,
  "chainId": undefined,
  "data": "0x608060405234801561000f575f80fd5b506040516101ef3803806101ef83398181016040528101906100319190610074565b805f819055505061009f565b5f80fd5b5f819050919050565b61005381610041565b811461005d575f80fd5b50565b5f8151905061006e8161004a565b92915050565b5f602082840312156100895761008861003d565b5b5f61009684828501610060565b91505092915050565b610143806100ac5f395ff3fe608060405234801561000f575f80fd5b5060043610610034575f3560e01c806360fe47b1146100385780636d4ce63c14610054575b5f80fd5b610052600480360381019061004d91906100ba565b610072565b005b61005c61007b565b60405161006991906100f4565b60405180910390f35b805f8190555050565b5f8054905090565b5f80fd5b5f819050919050565b61009981610087565b81146100a3575f80fd5b50565b5f813590506100b481610090565b92915050565b5f602082840312156100cf576100ce610083565b5b5f6100dc848285016100a6565b91505092915050565b6100ee81610087565b82525050565b5f6020820190506101075f8301846100e5565b9291505056fea264697066735822122019e943356c89506511b952171a3b4724d3152e5f4029bbb0ecc836d1365fcce464736f6c6343000816003300000000000000000000000000000000000000000000000000000000000001a4",
  "from": "0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266",
  "gas": 162455n,
  "gasPrice": 7n,
  "hash": "0x2a872fc2c05d90cbbdfbed7a5c831533dc1d02c1be4ab374b7d9c66e9ccec0e8",
  "maxFeePerGas": 7n,
  "maxPriorityFeePerGas": 0n,
  "nonce": 0,
  "to": null,
  "transactionIndex": 0,
  "type": "eip1559",
  "typeHex": "0x2",
  "v": undefined,
  "value": 0n,
  "yParity": undefined,
}
`;

exports[`viemPublicActions getTransactionReceipt should work 1`] = `
{
  "blockNumber": 1n,
  "contractAddress": "0x5fbdb2315678afecb367f032d93f642f64180aa3",
  "cumulativeGasUsed": 147721n,
  "effectiveGasPrice": null,
  "from": "0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266",
  "gasUsed": 147721n,
  "logs": [],
  "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  "status": null,
  "to": "0x",
  "transactionHash": "0x2a872fc2c05d90cbbdfbed7a5c831533dc1d02c1be4ab374b7d9c66e9ccec0e8",
  "transactionIndex": 0,
  "type": null,
}
`;

exports[`viemPublicActions simulateContract should work 1`] = `
{
  "request": {
    "abi": [
      {
        "inputs": [
          {
            "name": "x",
            "type": "uint256",
          },
        ],
        "name": "set",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function",
      },
    ],
    "account": undefined,
    "address": "0x5FbDB2315678afecb367f032d93F642f64180aa3",
    "args": [
      99999n,
    ],
    "bytecode": "0x608060405234801561000f575f80fd5b506040516101ef3803806101ef83398181016040528101906100319190610074565b805f819055505061009f565b5f80fd5b5f819050919050565b61005381610041565b811461005d575f80fd5b50565b5f8151905061006e8161004a565b92915050565b5f602082840312156100895761008861003d565b5b5f61009684828501610060565b91505092915050565b610143806100ac5f395ff3fe608060405234801561000f575f80fd5b5060043610610034575f3560e01c806360fe47b1146100385780636d4ce63c14610054575b5f80fd5b610052600480360381019061004d91906100ba565b610072565b005b61005c61007b565b60405161006991906100f4565b60405180910390f35b805f8190555050565b5f8054905090565b5f80fd5b5f819050919050565b61009981610087565b81146100a3575f80fd5b50565b5f813590506100b481610090565b92915050565b5f602082840312156100cf576100ce610083565b5b5f6100dc848285016100a6565b91505092915050565b6100ee81610087565b82525050565b5f6020820190506101075f8301846100e5565b9291505056fea264697066735822122019e943356c89506511b952171a3b4724d3152e5f4029bbb0ecc836d1365fcce464736f6c63430008160033",
    "dataSuffix": undefined,
    "deployedBytecode": "0x608060405234801561000f575f80fd5b5060043610610034575f3560e01c806360fe47b1146100385780636d4ce63c14610054575b5f80fd5b610052600480360381019061004d91906100ba565b610072565b005b61005c61007b565b60405161006991906100f4565b60405180910390f35b805f8190555050565b5f8054905090565b5f80fd5b5f819050919050565b61009981610087565b81146100a3575f80fd5b50565b5f813590506100b481610090565b92915050565b5f602082840312156100cf576100ce610083565b5b5f6100dc848285016100a6565b91505092915050565b6100ee81610087565b82525050565b5f6020820190506101075f8301846100e5565b9291505056fea264697066735822122019e943356c89506511b952171a3b4724d3152e5f4029bbb0ecc836d1365fcce464736f6c63430008160033",
    "functionName": "set",
    "humanReadableAbi": [
      "function set(uint256 x)",
    ],
    "to": "0x5FbDB2315678afecb367f032d93F642f64180aa3",
  },
  "result": undefined,
}
`;
